### ISSUE TITLE
[Help] Example of Apple SignIn implementation with POST_FORM

### ISSUE THREAD

**Issue by Lith** [simov/grant#193]
Hi 😄 ,

I've would like to implement Apple SignIn authentication with Grant + Purest, but I'm unable to find great configuration with Purest (not update since 4 years with latest providers like 'Apple')

`Grant` is unable to read POST_FORM `access_token` and it's return `oAuth2 missing code parameter` error.

Grant config (not sure about this)
```js
 apple: {
      key: '', // reverse.domain.name
      secret: '', // Need to be refer with "Generate client secret" 
      callback: `[domain.name]/auth/apple/callback`,
      nonce: true,
      state: true,
      scope: ['name', 'email'],
      custom_params: {
        response_type: 'code id_token',
        response_mode: 'form_post' // necessary... Apple constrain 😢 
      },
    },
```

Purest configuration testing :
```js
const apple = purest({
        provider: 'apple',
        config: {
          apple: {
            'https://appleid.apple.com': {
              __domain: {
                auth: {
                  auth: { bearer: '[0]' },
                },
              },
              '{endpoint}': {
                __path: {
                  alias: '__default',
                },
              },
            },
          },
        }
      });

apple
        .query()
        .post('auth/token') // not sure about this ? apple want post request only
        .auth(access_token)
        .request((err, res, body) => {  
              if (err) {
                   callback(err);
              } else {
                  // let do some stuff...
                  callback(null, {body});
              }
        });

```

I've successfully sign in with other libraries (`apple-signin` and `apple-auth`) but I would like to perform signIn request only with Grant + Purest and not implement a tierce package.
But `Grant` return the same error with the `access_token` (`code` for Apple) and it's unable to return and read it from POST_FORM request. 

The project "Playground" implement Apple SignIn testing example, but I can't find any code to understand how it's working.
Apple documentation is clearly... pitiable.

Thanks for your help !

**Comment by simov** [comment]
Hi @Lith, lets get Grant working first. As far as I understand you are not able to get an access_token for Apple? Can you check this https://github.com/simov/grant/pull/167#issuecomment-638292802 and see if it works for you. You have to generate the client secret in order to login.

Let me know if it works for you.

**Comment by Lith** [comment]
Hi @simov, 

Yes, I've implemented this function, but it's require in step2 (after called `auth/authorize` that return the `code`).
The `code` return by Apple, is not read by Grant because it's send in POST in Form-Data and not in GET response params.

Not sure if I'm clear about that... 
But the generated token is require after, and this process is working well.

Which default configuration are you using for Grant with Apple ?
Are you using Purest or an other library ?

Thanks :)

**Comment by simov** [comment]
I see, seems like I never got around implementing it in Grant .. it should be an easy fix, but you'll have to wait a little bit, I'll let you know when it's done.

As for the Purest config, you shouldn't be doing the last request of the OAuth flow manually, that's what this module is for. Otherwise your Purest config looks ok.

**Comment by simov** [comment]
Published the fix in `v5.4.4`, however in case you are using Strapi you have to go there and ask them to bump the dependency version, because it seems that they have pinned it down to [5.2.0](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/package.json#L17) for some reason, and they won't receive the patch automatically.

Let me know if it works for you.

**Comment by Lith** [comment]
Thank you @simov ! I will test it right now 👍 

I'm using a "custom" version of Strapi, I will be able to update it on latest release :)

**Comment by simov** [comment]
Let me know if you have any other issues with this.

**Comment by EliSadaka** [comment]
I'm also having issues implementing Apple sign-in with Grant. This is what my config looks like:

```
      "defaults": {
        "origin": process.env.SERVER_URL,
        "prefix": "/auth",
        "transport": "querystring",
        "response": ["tokens"],
        "state": true,
        "nonce": true
      },
      "apple": {
        "key": " ... ",
        "secret": generateAppleClientSecret(),
        "scope": ["name", "email"],
        "custom_params": {
          "response_type": "code id_token",
          "response_mode": "form_post"
        }
      }
```

`generateAppleClientSecret()` returns the token generated by this function: https://github.com/simov/grant/pull/167#issuecomment-638292802

The `/auth/apple/callback` endpoint is hit with a POST request containing the response data in the body, but then it redirects the client to root with the `missing session or misconfigured provider` error. I tried adding an Express middleware function to the `/auth/apple/callback` path that moves the response data to the query params and changes the method to GET, but it didn't seem to make a difference (it's possible that Grant is handling the request before the middleware function does). I have Google and Facebook sign-in working so I think the issue is specific to my Apple provider configuration.

Any help is appreciated.

**Comment by simov** [comment]
Do you have a body-parser middleware mounted before Grant?

**Comment by EliSadaka** [comment]
> 
> 
> Do you have a body-parser middleware mounted before Grant?

Yes, I do:

```
app.use(express.json());
app.use(express.urlencoded({ extended: true }));
```

**Comment by simov** [comment]
Make sure you are using latest version of Grant. Other than that I can't test with Apple in particular because I don't have a developer account there. However I tested with [microsoft](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow), because they support that parameter, and it works. Another thing to try out is to add your config and use the default [example](https://github.com/simov/grant/blob/master/examples/transport-querystring/express.js).

**Comment by rxb** [comment]
@EliSadaka Were you able to resolve this? I'm running into the same issue, I think.

**Comment by EliSadaka** [comment]
@rxb No, I wasn't able to resolve my issue so I went about implementing Apple sign-in in a completely different way.

**Comment by rxb** [comment]
Digging into it more, I don't think it's possible for Grant to successfully OAuth with Apple without some serious custom code due to Apple's extremely weird OAuth implementation. `form_post` works just fine, but the issue seems to be that the user data is already included in the body of the initial OAuth response. There is no access token and there is no secondary request to get the user/profile resource. What's more, the user data is packaged up in a signed jwt that must be decoded/verified using one of 3 public JSON Web Keys hosted on an Apple server... and you must snag a value from a header of the original OAuth response to know which of the 3 keys to use.

This is all to day, Apple's implementation of OAuth is extremely nuts and at this point I think I have to admit that I would be in way over my head to try to make a pull request to work it into Grant.

**Comment by rxb** [comment]
Digging into it more, I don't think it's possible for Grant to successfully OAuth with Apple without some serious custom code due to Apple's extremely weird OAuth implementation. form_post works just fine, but the issue seems to be that the user data is already included in the body of the initial OAuth response. There is no access token and there is no secondary request to get the user/profile resource. What's more, the user data is packaged up in a signed jwt that must be decoded/verified using one of 3 public JSON Web Keys hosted on an Apple server... and you must snag a value from a header of the original OAuth response to know which of the 3 keys to use.

**Comment by simov** [comment]
I see. I'm just wondering how these guys over here https://github.com/simov/grant/pull/167 were able to get the returned `id_token` .. 

In any case, I can implement this and push it to a branch if you are willing to test it on your end, since I don't have a developer account on Apple.

Let me know.

**Comment by rxb** [comment]
Thanks so much. I am now second guessing myself after seeing that is seemed to work for #167. It's very possible I'm not doing something correctly. I'm going to set up a much simpler project to really see if I can isolate what's going on.

**Comment by simov** [comment]
You can use any of the [basic examples](https://github.com/simov/grant/tree/master/examples/transport-session).

The `id_token` is supposed to be returned by the second request, so in case it's being returned by the first one isntead I don't see how they got it working without patching Grant.

**Comment by rxb** [comment]
Did a basic example (https://github.com/rxb/grant/tree/master/examples/transport-session) and got back the whole id_token, so it looks like I must be way off in my understanding of how Apple's OAuth is working. The issue must be something else (I'm using Grant indirectly through FeathersJs), so I'll have to keep digging. Thanks so much for your help, I really appreciate it!

**Comment by simov** [comment]
I got mislead by [this document](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms). The example there is using `response_type: id_token` where in Grant you have `response_type: code` instead, so I missed that part and I was looking at the response data only.

I'm glad it worked for you in the end :+1:

**Comment by forgot** [comment]
@simov I've been digging into this a little more, so I thought I'd add what I've found. Like @rxb, I'm using Grant through indirectly through FeathersJS, and I'm trying to figure out exactly where the problem is. My `config` is the same as above, as well as the function used to generate the secret.

There is a successful `POST` to `/callback`, but the Form Data returned isn't the same as the response data.

Here's an example of the Form Data sent as a `POST` to `/callback`:
```
state: some_state_code
code: some_code
id_token: some_id_token
user: {"name": {"firstName":"Homer","lastName":"Simpson"},"email":"homer@simpson.com"}
```

Here is the response data:
```
{
  strategy: 'some_strategy',
  id_token: 'some_id_token',
  access_token: 'some_access_token',
  refresh_token: 'some_refresh_token',
  raw: {
    access_token: 'some_access_token',
    token_type: 'Bearer',
    expires_in: 3600,
    refresh_token: 'some_refresh_token',
    id_token: 'some_id_token'
  },
  profile: { error: 'Grant: No profile URL found!' }
}
```

Everything but the `id_token` is lost. I'm not sure if that's a Grant implementation thing, or a FeathersJS one though...

**Comment by simov** [comment]
Support for this feature landed in Grant [v5.4.4 (2020/09/23)](https://github.com/simov/grant/blob/master/CHANGELOG.md#v544-20200923)

The current stable release of Feathers `v4.5.11` is still using Grant [v4.7.0](https://github.com/feathersjs/feathers/blob/crow/packages/authentication-oauth/package.json#L61)

Where the upcoming release `v5.0.0-pre.3` is using [v5.4.13](https://github.com/feathersjs/feathers/blob/dove/packages/authentication-oauth/package.json#L61) instead.

Can you check which version of Feathers do you use?

**Comment by forgot** [comment]
Sorry, should have mentioned that I'm using Feathers `5.0.0-pre.3` with Grant `v5.4.13`.

Like I said, I'm not sure where this info is being inadvertently stripped, but I plan on doing more digging today.

**Comment by simov** [comment]
I'm kind of lost. Are you receiving the correct response data or not? What exactly is being stripped? The `user` field? In case you are receiving the access, refresh and id token, then that is correct.

**Comment by forgot** [comment]
Did a little more digging, and I think I *might* know what's happening. Grant is correctly returning the info after verification of the identity token, but some of the information that is returned as part of the authorization step is missing, and it's needed to create the user. Apple's documentation on this is, to me, extremely confusing, and I've been reading Apple documentation for years. For the sake of clarity, I'll break down step-by-step what I see happening.

#### Authorization
- The user clicks on a link to kick off OAuth (ex: `/oauth/apple`)
- The user is redirected to the `authorize_url`, which in this case is `https://appleid.apple.com/auth/authorize`
  - The following query string parameters are attached:
      ```
      client_id = com.company.awesomesite.client
      response_type = code token
      redirect_uri = https://awesomesite.com/oauth/apple/callback
      scope = name email
      state = (some generated state)
      nonce = (some generated nonce)
      response_mode = form_post
      ```
- After entering password (and 2FA if enabled), the user is redirected to `https://appleid.apple.com/appleauth/auth/oauth/consent`
  - The same query string parameters listed above are attached
  - The user is chooses which information to share, and clicks "Continue" to allow 
- After a few internal redirects, Apple sends a `POST` to the `redirect_uri` (in this example `https://awesomesite.com/oauth/apple/callback`)
  - Since the `response_mode` is `form_post`, Apple sends the following Form Data:
    ```
    state: (same value sent as query string parameter above)
    code: (a single use authorization code)
    id_token: (a JWT)
    user: {"name":{"firstName":"Homer","lastName":"Simpson"},"email":"homer@simpson.com"}
    ```
  - The decoded `id_token` contains the following:
    ```
    {
      "kid": "(some key id)",
      "alg": "RS256"
    },
    {
      "iss": "https://appleid.apple.com",
      "aud": "com.company.awesomesite.client",
      "exp": 1620407571,
      "iat": 1620321171,
      "sub": "(some subject data)",
      "nonce": "(same value sent as query string parameter above)",
      "c_hash": "(hashed code sent with this JWT)",
      "email": "homer@simpson.com",
      "email_verified": "true",
      "is_private_email": "true",
      "auth_time": 1620321170,
      "nonce_supported": true
    }
    ```

#### Verification
Here is where things get a little confusing to me. I believe Grant uses the above information to successfully:
- [Generate and Validate the token](https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens)
- Return a [TokenResponse](https://developer.apple.com/documentation/sign_in_with_apple/tokenresponse)

#### The Problem
After verification, we end up with the [TokenResponse](https://developer.apple.com/documentation/sign_in_with_apple/tokenresponse) information only. While the `id_token` in that response has some of the information needed, any `user` information that was included in the original Form Data sent when Apple sent a `POST` to the `redirect_uri` is gone. That Form Data sent via `POST` to the `redirect_uri` is the only time Apple sends a `user` object, and some of that data is not included in later requests. This `user` information, along with information contained in the `id_token`, is used to create the user on the server.

#### The Question
Is it possible to hold on to that `user` information sent in the Form Data of the `POST` to `redirect_uri` after the verification? This may be a trivial thing in Grant, and may be an implementation issue of FeathersJS, but I want to be sure. Using the FeathersJS implementation of Grant, the data that gets returned looks like this:
```
{
  strategy: 'apple',
  id_token: '(some id token)',
  access_token: '(some access token)',
  refresh_token: '(some refresh token)',
  raw: {
    access_token: '(some access token)',
    token_type: 'Bearer',
    expires_in: 3600,
    refresh_token: '(some refresh token)',
    id_token: '(some id token)'
  },
  profile: { error: 'Grant: No profile URL found!' }
}
```

Is there a way that it could be returned in a way similar to this?
```
{
  strategy: 'apple',
  id_token: '(some id token)',
  access_token: '(some access token)',
  refresh_token: '(some refresh token)',
  raw: {
    access_token: '(some access token)',
    token_type: 'Bearer',
    expires_in: 3600,
    refresh_token: '(some refresh token)',
    id_token: '(some id token)'
  },
  profile: {'name':{'firstName':'Homer','lastName':'Simpson'},'email':'homer@simpson.com'}
}
```

**Comment by simov** [comment]
Got it, thanks for putting this up, I really appreciate it.

According to the [OpenID Configuration](https://appleid.apple.com/.well-known/openid-configuration) endpoint Apple supports `openid`, `email` and `name` scopes. Can you verify that after including all those 3 scopes the `id_token` still contains only the email, but not the user name.

If that's the case then this needs to be implemented in Grant. That user data returned after the first call is not expected to be there when using the `response_type: code`. It will be fairly trivial to implement here, but you will have to test it because I don't have a developer account on Apple.

Another thing missing inside the OpedID Configuration endpoint is a `userinfo_endpoint` which otherwise could've been used for obtaining the user profile after the authorization flow.

**Comment by forgot** [comment]
I tried using all three scopes, but it did not change the content of the `id_token`. All of the "Sign in with Apple" documentation says the only allowed scopes are `name` and `email`. I wonder if the configuration link you sent is part of one of their social logins?

The `user` object that is returned after the first call is absolutely not a standard spec. There's a callout at the link [here](https://sarunw.com/posts/sign-in-with-apple-4/#handle-the-redirect) that talks about that a little more. Also, I haven't read anything at all about a `userinfo_endpoint` in any of the "Sign in with Apple" documentation.

**Comment by rxb** [comment]
Not sure if this will be useful to your use case @forgot or maybe others who find their way to this thread, but I threw together a short guide to using Apple Oauth with FeathersJS. I had a lot of misconceptions going into it, but once I got myself sorted out, this is what I ended up with:
https://gist.github.com/rxb/e596c66b03e3262f26d9ede5d7dbab81

**Comment by forgot** [comment]
A couple of points to add...

 In addition to a [TokenResponse](https://developer.apple.com/documentation/sign_in_with_apple/tokenresponse), an [ErrorResponse](https://developer.apple.com/documentation/sign_in_with_apple/errorresponse) can also be returned. I recommend reading the [Generate and Validate Tokens](https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens) documentation if you haven't already.

Also, according to Apple:
>To verify the identity token, your app server must:
>  - Verify the JWS E256 signature using the server’s public key
>  - Verify the nonce for the authentication
>  - Verify that the iss field contains https://appleid.apple.com
>  - Verify that the aud field is the developer’s client_id
>  - Verify that the time is earlier than the exp value of the token

[Here](https://github.com/simov/grant#connect-openid-connect) the Grant documentation states that Grant **does not** verify the signature of the returned `id_token` by default, but does validate the `aud` and `nonce` claims.

  My question here is, if some of this is already being handled automatically, where can this be overridden or extended to follow Apple's recommendations? (This could also be done after Grant has completed it's default validation and returned since all the necessary information is already in the token itself.)

- Digging deeper into validation, to verify the `JWS E256` signature requires an additional endpoint of `https://appleid.apple.com/auth/keys`, which is returns a variable array of [JWKSet.Keys](https://developer.apple.com/documentation/sign_in_with_apple/jwkset/keys). You can read more [here](https://developer.apple.com/documentation/sign_in_with_apple/fetch_apple_s_public_key_for_verifying_token_signature).

I really appreciate that you're taking the time to walk through this with me, and I'm happy to do whatever I can to help test things out.

**Comment by simov** [comment]
The OpenID Connect protocol requires from the identity providers to have a publicly accessible [configuration endpoint](https://appleid.apple.com/.well-known/openid-configuration) that outlines what that provider supports. There the `jwks_uri` endpoint contains the public keys that you can use to verify your `id_token`.

Initially that was implemented as a [separate module](https://github.com/simov/grant-oidc/blob/master/lib/id_token.js) but now my plan is to integrate it into Grant. But for the time being you have to validate it yourself. You can use the `jws` module to decode and validate the token.

The additional data that Apple returns when redirecting you back to your server is not expected. The OpenID Connect protocol defines a `userinfo_endpoint` that can be used to obtain the user profile after successful login.

I'm going to patch Grant to persist that data in session and populate the `profile` response key. I will push the patch into a branch so that you, @forgot, can pull it and validate it. I will provide more details when I'm ready. After that I'm going to add tests and release it as a patch.

Validating the `id_token` in Grant will take a bit longer, but it's the next item in my backlog.

**Comment by simov** [comment]
Hi, @forgot, I pushed the fix to master https://github.com/simov/grant/commit/15d64bd913474764a47ee4a0e5fce10dc55e1b6c so all you need to do is pull Grant from GitHub:

```
"grant": "simov/grant",
```

You need to specify the `response: [..., 'profile']` option as usual.

Let me know if it works for you.

Also can you test this method for generating the client secret:

```js
var jws = require('jws')

exports.sign = ({client_id, key_id, team_id, private_key}) => {
  var epoch = Math.floor(new Date().getTime() / 1000)

  var options = {
    header: {
      typ: 'JWT',
      alg: 'ES256',
      kid: key_id,
    },
    payload: {
      iss: team_id,
      sub: client_id,
      aud: 'https://appleid.apple.com',
      iat: epoch,
      exp: epoch + (60 * 60 * 24), // 1 day
    },
    secret: private_key
  }

  return jws.sign(options)
}
```

I'm asking because Grant uses `jws` internally and I would have used that instead of `jsonwebtoken`. Note that `jsonwebtoken` is using `jws` under the hood too, but there might be some slight adjustments needed. The `private_key` have to be string read using `fs`.

**Comment by forgot** [comment]
Sorry for the delay here. I can tell you that

**Comment by forgot** [comment]
Ok, the commit at 15d64bd and the method to generate the client secret using `jws` both work! Like I said before, I'm using Grant indirectly through FeathersJS, but everything works smoothly.

Thank you for your work on this! I know there's still a little to be done to validate the `id_token` to Apple's specifications, so please let me know if I can be of assistance. Once it's wrapped up and bumped, I'll see about getting it pulled into Feathers as well.

**Comment by simov** [comment]
That's great news, I'll publish the patch for the user data over the weekend. As mentioned earlier validating the `id_token` will take a bit longer to implement because it's not just about Apple, but having that feature in general for any provider supporting OpenID Connect.

Automatically generating the client secret for Apple won't be part of this patch either, because I don't have a good way to pass those additional parameters, like `team_id`.

I'll let you know when the patch is released.

**Comment by simov** [comment]
The fix for the user profile was published in v5.4.14

**Comment by EloB** [comment]
I've tried to implement this into Strapi. I never managed get the `user` information only `email`. Do anyone have a clue way?

https://github.com/strapi/strapi/pull/15522
